File: scalademo/src/com/milko/trainings/scala/generics/scalac/ListStack.scala
package com.milko.trainings.scala.generics.scalac

class ListStack[T] extends Stack[T] {
  
  private var elements: List[T] = List();
  
  override def pop: T = {
    if (elements.isEmpty)
      throw new IndexOutOfBoundsException()
      
    val head = elements.head;
    elements = elements.tail;
    head;
  }

  def push(t: T): Unit = {
    elements = t+:elements;
  }
}

  
  object ListStack {
   def main(args: Array[String]) : Unit = {
    val stack = new ListStack[String];
    stack.push("C");
    stack.push("B");
    stack.push("A");
    
    println(stack.pop)
    println(stack.pop)
    println(stack.pop)
    
    
    }
  }
  

File: scalademo/src/com/milko/trainings/scala/generics/javac/ListStack.java
package com.milko.trainings.scala.generics.javac;

import java.util.ArrayList;
import java.util.List;

public class ListStack<T> implements Stack<T> {

	private final List<T> elements = new ArrayList<>();
	
	@Override
	public void push(T t) {
		elements.add(0, t);
		
	}

	@Override
	public T pop() {
		if (elements.isEmpty())
			throw new IndexOutOfBoundsException();
		return elements.remove(0);
	}

	public static void main(String[] args) {
		Stack<String> stack = new ListStack();
		
		stack.push("C");
		stack.push("B");
		stack.push("A");
		
		//String element = stack.pop();
		System.out.println(stack.pop());
		System.out.println(stack.pop());
		System.out.println(stack.pop());
		
		
		
		
	}
	
}
